### PR TITLE
editor: add schema example in the demo application

### LINKED ISSUE #115: Create an example of the editor with a local JSONSchema
**Describe the issue**

Add an example to demonstrate the editor capabilities to the ng-core-tester.

### PR DESCRIPTION
**Pull request by jma**
* Adds a new link in the application demo to show the editor with a
  specific JSONSchema to demonstrate all the possibilities of the editor.
* Makes JSONSchema and editorSettings as Input for the editor component.
* Cleans the empty values in the emitted model.
* Set the defauls model and JSONSchema values if the URL do not contains
  the record type.
* Closes #115.

Co-Authored-by: Johnny Mariéthoz <Johnny.Mariethoz@rero.ch>

## Why are you opening this PR?

- To diagnose easily the editor.

## How to test?

- Start the demo application and go to the record menu and click on the editor entry.

__Note__: the editor component file has been re-indented. Sorry for the review.

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Extracted translations?

### CODE REVIEW

File: projects/ng-core-tester/src/app/record/editor/editor.component.html [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<div class="card mb-3">
+  <div class="card-header">
+    Model
+  </div>
+  <div class="card-body">
+    <pre>{{model | json}}</pre>
```
Comment (Garfield-fr) [inline projects/ng-core-tester/src/app/record/editor/editor.component.html:6]: space after `{`and before `}`

**Review by Garfield-fr — APPROVED** [review]

File: projects/ng-core-tester/src/app/record/editor/schema.json [hunk 2]
```diff
@@ -0,0 +1,396 @@
+{
+  "schema": "http://json-schema.org/draft-07/schema#",
+  "type": "object",
+  "title": "Test Editor",
```
Comment (iGormilhit) [inline projects/ng-core-tester/src/app/record/editor/schema.json:4]:
```suggestion
  "title": "Test editor",
```

File: projects/ng-core-tester/src/app/record/editor/schema.json [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+    "required"
+  ],
+  "properties": {
+    "oneOf": {
+      "title": "oneOf",
+      "type": "object",
+      "additionnalProperties": "false",
+      "oneOf": [
+        {
+          "title": "prop1",
```
Comment (iGormilhit) [inline projects/ng-core-tester/src/app/record/editor/schema.json:30]:
```suggestion
          "title": "Proposition 1",
```

File: projects/ng-core-tester/src/app/record/editor/schema.json [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+            "name",
+            "type",
+            "hidden"
+          ],
+          "required": [
+            "name"
+          ],
+          "properties": {
+            "name": {
+              "title": "name",
```
Comment (iGormilhit) [inline projects/ng-core-tester/src/app/record/editor/schema.json:41]:
```suggestion
              "title": "Name",
```

File: projects/ng-core-tester/src/app/record/editor/schema.json [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+              "type": "string",
+              "form": {
+                "templateOptions": {
+                  "cssClass": "col-lg-6"
+                }
+              }
+            },
+            "hidden": {
+              "title": "hidden name",
```
Comment (iGormilhit) [inline projects/ng-core-tester/src/app/record/editor/schema.json:50]:
```suggestion
              "title": "Hidden name",
```

File: projects/ng-core-tester/src/app/record/editor/schema.json [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+              "type": "string",
+              "form": {
+                "hide": true
+              },
+              "templateOptions": {
+                "cssClass": "col-lg-6"
+              }
+            },
+            "type": {
+              "title": "type",
```
Comment (iGormilhit) [inline projects/ng-core-tester/src/app/record/editor/schema.json:60]:
```suggestion
              "title": "Type",
```

File: projects/ng-core-tester/src/app/record/editor/schema.json [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+            }
+          },
+          "form": {
+            "templateOptions": {
+              "cssClass": "row"
+            }
+          }
+        },
+        {
+          "title": "prop2",
```
Comment (iGormilhit) [inline projects/ng-core-tester/src/app/record/editor/schema.json:79]:
```suggestion
          "title": "Proposition 2",
```

File: projects/ng-core-tester/src/app/record/editor/schema.json [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+          "required": [
+            "name"
+          ],
+          "properties": {
+            "name": {
+              "title": "name",
```
Comment (iGormilhit) [inline projects/ng-core-tester/src/app/record/editor/schema.json:85]:
```suggestion
              "title": "Name",
```

File: projects/ng-core-tester/src/app/record/editor/schema.json [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+              "type": "string"
+            },
+            "type": {
+              "title": "type",
```
Comment (iGormilhit) [inline projects/ng-core-tester/src/app/record/editor/schema.json:89]:
```suggestion
              "title": "Type",
```

File: projects/ng-core-tester/src/app/record/editor/schema.json [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+      ],
+      "form": {
+        "templateOptions": {
+          "cssClass": "editor-title"
+        }
+      }
+    },
+    "alwaysHidden": {
+      "title": "Always Hidden",
+      "description": "not present in the form, not present in the model",
```
Comment (iGormilhit) [inline projects/ng-core-tester/src/app/record/editor/schema.json:111]:
```suggestion
      "title": "Always hidden",
      "description": "Not present in the form, not present in the model.",
```

File: projects/ng-core-tester/src/app/record/editor/schema.json [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+      "title": "Hidden",
+      "type": "string",
+      "minLength": 1,
+      "form": {
+        "hide": true
+      }
+    },
+    "defaultHidden": {
+      "title": "Default Hidden",
+      "description": "Hidden Field with Default Value",
```
Comment (iGormilhit) [inline projects/ng-core-tester/src/app/record/editor/schema.json:126]:
```suggestion
      "title": "Default hidden",
      "description": "Hidden field with default value.",
```

File: projects/ng-core-tester/src/app/record/editor/schema.json [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+            "value": "value 1",
+            "label": "val1"
+          },
+          {
+            "value": "value 2",
+            "label": "val2"
+          },
+          {
+            "value": "value 3",
+            "label": "val3"
```
Comment (iGormilhit) [inline projects/ng-core-tester/src/app/record/editor/schema.json:184]:
```suggestion
            "value": "Value 1",
            "label": "val1"
          },
          {
            "value": "Value 2",
            "label": "val2"
          },
          {
            "value": "Value 3",
            "label": "val3"
```

File: projects/ng-core-tester/src/app/record/editor/schema.json [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+        }
+      },
+      "form": {
+        "templateOptions": {
+          "cssClass": "editor-title"
+        }
+      }
+    },
+    "object_of_object": {
+      "title": "Object of Object",
```
Comment (iGormilhit) [inline projects/ng-core-tester/src/app/record/editor/schema.json:217]:
```suggestion
      "title": "Object of object",
```

File: projects/ng-core-tester/src/app/record/editor/schema.json [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+      "type": "object",
+      "additionalProperties": false,
+      "propertiesOrder": [
+        "object1",
+        "prop3"
+      ],
+      "properties": {
+        "object1": {
+          "title": "Object1",
```
Comment (iGormilhit) [inline projects/ng-core-tester/src/app/record/editor/schema.json:226]:
```suggestion
          "title": "Object 1",
```

File: projects/ng-core-tester/src/app/record/editor/schema.json [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+          "cssClass": "editor-title"
+        }
+      }
+    },
+    "array": {
+      "title": "Array",
+      "type": "array",
+      "minItems": 1,
+      "items": {
+        "title": "properties",
```
Comment (iGormilhit) [inline projects/ng-core-tester/src/app/record/editor/schema.json:263]:
```suggestion
        "title": "Properties",
```

File: projects/ng-core-tester/src/app/record/editor/schema.json [hunk 16]
```diff
@@ -0,0 +1,396 @@
...
+        "propertiesOrder": [
+          "prop1",
+          "prop2"
+        ],
+        "required": [
+          "prop1"
+        ],
+        "properties": {
+          "prop1": {
+            "title": "Prop 1",
```
Comment (iGormilhit) [inline projects/ng-core-tester/src/app/record/editor/schema.json:275]:
```suggestion
            "title": "Proposition 1",
```

File: projects/ng-core-tester/src/app/record/editor/schema.json [hunk 17]
```diff
@@ -0,0 +1,396 @@
...
+            "type": "string",
+            "form": {
+              "templateOptions": {
+                "cssClass": "col"
+              }
+            }
+          },
+          "prop2": {
+            "title": "Prop 2",
```
Comment (iGormilhit) [inline projects/ng-core-tester/src/app/record/editor/schema.json:284]:
```suggestion
            "title": "Proposition 2",
```

File: projects/ng-core-tester/src/app/record/editor/schema.json [hunk 18]
```diff
@@ -0,0 +1,396 @@
...
+        }
+      },
+      "form": {
+        "templateOptions": {
+          "cssClass": "editor-title"
+        }
+      }
+    },
+    "array_of_array": {
+      "title": "Array Of Array",
```
Comment (iGormilhit) [inline projects/ng-core-tester/src/app/record/editor/schema.json:307]:
```suggestion
      "title": "Array of array",
```

File: projects/ng-core-tester/src/app/record/editor/schema.json [hunk 19]
```diff
@@ -0,0 +1,396 @@
...
+      "default": [
+        [
+          "value 1",
+          "value 2"
+        ],
+        [
+          "value 3"
```
Comment (iGormilhit) [inline projects/ng-core-tester/src/app/record/editor/schema.json:314]: I'm not sure here, should it have uppercases?
```suggestion
          "Value 1",
          "Value 2"
        ],
        [
          "Value 3"
```

File: projects/ng-core-tester/src/app/record/editor/schema.json [hunk 20]
```diff
@@ -0,0 +1,396 @@
...
+        "templateOptions": {
+          "cssClass": "editor-title"
+        },
+        "navigation": {
+          "essential": true
+        }
+      }
+    },
+    "array_of_object_of_object": {
+      "title": "Array of Object of Object",
```
Comment (iGormilhit) [inline projects/ng-core-tester/src/app/record/editor/schema.json:344]:
```suggestion
      "title": "Array of object of object",
```

File: projects/ng-core-tester/src/app/record/editor/schema.json [hunk 21]
```diff
@@ -0,0 +1,396 @@
...
+      "type": "array",
+      "minItems": 0,
+      "items": {
+        "title": "Object of Object",
```
Comment (iGormilhit) [inline projects/ng-core-tester/src/app/record/editor/schema.json:348]:
```suggestion
        "title": "Object of object",
```

File: projects/ng-core-tester/src/app/record/editor/schema.json [hunk 22]
```diff
@@ -0,0 +1,396 @@
...
+        "type": "object",
+        "additionalProperties": false,
+        "propertiesOrder": [
+          "object1",
+          "prop3"
+        ],
+        "properties": {
+          "object1": {
+            "title": "Object1",
```
Comment (iGormilhit) [inline projects/ng-core-tester/src/app/record/editor/schema.json:357]:
```suggestion
            "title": "Object 1",
```

**Review by iGormilhit — COMMENTED** [review]
Commit message is fine.

Otherwise, it's juste a matter of coherence in the JSON schema.

File: projects/ng-core-tester/src/app/record/editor/editor.component.html [hunk 23]
```diff
@@ -0,0 +1,10 @@
+<div class="card mb-3">
```
Comment (sebastiendeleze) [inline projects/ng-core-tester/src/app/record/editor/editor.component.html:1]: Header is missing

File: projects/ng-core-tester/src/app/record/editor/editor.component.spec.ts [hunk 24]
```diff
@@ -0,0 +1,37 @@
+import { HttpClientTestingModule } from '@angular/common/http/testing';
```
Comment (sebastiendeleze) [inline projects/ng-core-tester/src/app/record/editor/editor.component.spec.ts:1]: Header is missing

File: projects/ng-core-tester/src/app/record/editor/editor.component.ts [hunk 25]
```diff
@@ -0,0 +1,38 @@
+import { Component } from '@angular/core';
```
Comment (sebastiendeleze) [inline projects/ng-core-tester/src/app/record/editor/editor.component.ts:1]: Header is missing

File: projects/ng-core-tester/src/app/record/editor/editor.component.ts [hunk 26]
```diff
@@ -0,0 +1,38 @@
...
+import JSONSchema from './schema.json';
+@Component({
+  selector: 'app-editor',
+  templateUrl: './editor.component.html',
+  styleUrls: ['./editor.component.scss']
```
Comment (sebastiendeleze) [inline projects/ng-core-tester/src/app/record/editor/editor.component.ts:6]: You can remove `styleUrls` property as it's not used, and the `scss` file, too.

File: projects/rero/ng-core/src/lib/record/editor/editor.component.ts [hunk 27]
```diff
@@ -168,6 +169,9 @@ export class EditorComponent implements OnInit, OnChanges, OnDestroy {
         this.setSchema(this.schema);
       }
     }
+    if (changes.schema && changes.schema.isFirstChange()) { // Schema has changed
+      this.setSchema(changes.schema.currentValue);
+    }
```
Comment (sebastiendeleze) [inline projects/rero/ng-core/src/lib/record/editor/editor.component.ts:174]: Can be referenced in commit message

File: projects/rero/ng-core/src/lib/record/editor/editor.component.ts [hunk 28]
```diff
@@ -187,15 +191,20 @@ export class EditorComponent implements OnInit, OnChanges, OnDestroy {
         // );
         this._spinner.show();
         this.loadingChange.emit(true);
-
+        if (!params.type) {
+          this.model = {};
+          this.schema = {};
+          this._spinner.hide();
+          return;
+        }
```
Comment (sebastiendeleze) [inline projects/rero/ng-core/src/lib/record/editor/editor.component.ts:199]: Can be referenced in commit message

File: projects/rero/ng-core/src/lib/record/editor/editor.component.ts [hunk 29]
```diff
@@ -320,7 +329,7 @@ export class EditorComponent implements OnInit, OnChanges, OnDestroy {
    * @param modelValue Model.
    */
   modelChanged(modelValue: any) {
-    this.modelChange.emit(modelValue);
+    this.modelChange.emit(removeEmptyValues(modelValue));
```
Comment (sebastiendeleze) [inline projects/rero/ng-core/src/lib/record/editor/editor.component.ts:332]: Can be referenced in commit message

File: projects/ng-core-tester/src/app/record/editor/schema.json (hunk 19 above)
Comment (jma) [reply projects/ng-core-tester/src/app/record/editor/schema.json:314]: It is up to us. They are values.

File: projects/rero/ng-core/src/lib/record/editor/editor.component.ts (hunk 27 above)
Comment (jma) [reply projects/rero/ng-core/src/lib/record/editor/editor.component.ts:174]: It is actually as the schema has been added to inputs.

File: projects/rero/ng-core/src/lib/record/editor/editor.component.ts (hunk 29 above)
Comment (jma) [reply projects/rero/ng-core/src/lib/record/editor/editor.component.ts:332]: Done

File: projects/rero/ng-core/src/lib/record/editor/editor.component.ts (hunk 28 above)
Comment (jma) [reply projects/rero/ng-core/src/lib/record/editor/editor.component.ts:199]: Done

**Review by sebastiendeleze — APPROVED** [review]

**Review by iGormilhit — APPROVED** [review]

**Review by iGormilhit — CHANGES_REQUESTED** [review]
Mention https://github.com/rero/ng-core/issues/115 in the commit message please.

**Pull request merged**